### PR TITLE
Added early experimental SEV-SNP support

### PR DESCRIPTION
**Pull request by ihaller**
Starting this line of work to make sure it does not get lost in private branches.
Known limitation:
- No handler for #VC interrupt (it needs to be specially designed to be hardened) meaning that there are no IPIs so multi-threaded code can deadlock.
- Shared memory I/O not functioning well, need to have better understanding of how the loaders sets up the memory.

### CODE REVIEW

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+        current_entry_end_address > vmsa_settings->virtual_top_of_memory)
+      {
+        if (first_vtom_entry)
+        {
+          size_t above_vtom_size =
+            current_entry_end_address - vmsa_settings->virtual_top_of_memory;
+          above_vtom_memory_range = std::span(
+            reinterpret_cast<uint8_t*>(vmsa_settings->virtual_top_of_memory),
+            above_vtom_size);
+          current_entry_end_address -= above_vtom_size;
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc]: Is this equivalent to:
```suggestion
          current_entry_end_address = vmsa_settings->virtual_top_of_memory;
```

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+          entry_size -= above_vtom_size;
+          first_vtom_entry = false;
+        }
+        // Extra memory ranges beyond the top of the virtual top-of-memory are
+        // ignored.
+        else
+        {
+          entry_size = 0;
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc]: I wonder if
```suggestion
          continue;
```
would be clearer.  And you move
```C++
last_entry_end_address = current_entry_end_address;
```
a bit earlier, e.g. line 416.

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+      size_t current_entry_end_address = entry_address + entry_size;
+      if (
+        current_entry_end_address < entry_address ||
+        current_entry_end_address > (static_cast<uint64_t>(1) << 48))
+      {
+        kabort();
+      }
+      if (
+        first_entry &&
+        current_entry_end_address <= snmalloc::address_cast(heap_start))
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc]: Do we need to add ` || snmalloc::address_cast(heap_start) < entry_address`?

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+      (snmalloc::address_cast(heap_start) >=
+         vmsa_settings->virtual_top_of_memory ||
+       (vmsa_settings->virtual_top_of_memory % PAGE_SIZE) != 0))
+    {
+      kabort();
+    }
+
+    bool first_entry = true;
+    bool first_vtom_entry = true;
+    size_t last_entry_end_address = 0;
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc]: I don't know if you want to do this, but would it be clearer to use `address_t`?
```suggestion
    address_t last_entry_end_address = 0;
```

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+    add_to_kernel_pagetable(
+      EXTRA_DATA_RANGE.start, EXTRA_DATA_RANGE.size(), PT_KERNEL_READ);
+  }
+
+  /**
+   * Lazy acceptance of memory as the heap expands.
+   */
+  static void notify_using_memory_sev(std::span<uint8_t>)
+  {
+    // TODO: Implement lazy acceptance. Some ranges do not get notified about
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc:520]: Just flagging to check this is intentionally not part of this PR.

File: guest-verona-rt/arch/x86_64/hv_sev.cc [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+    // today.
+    // accept_private_memory(AddressRange(range));
+  }
+
+  static void shutdown_sev()
+  {
+    write_msr(SEV_MSR_GHCB, SevGhcbMsrTerminationRequest(0).raw());
+    vmgexit();
+    // Infinite loop since host can refuse termination request
+    while (true)
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/hv_sev.cc:530]: Do we have to worry about other threads if the host refuses?

File: guest-verona-rt/arch/x86_64/inc/sev_msr.h [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+   * Generic root implementation for request taking a 48-bit input.
+   * Combines the 12-bit info field with the input to form 64-bit value.
+   */
+  template<SevGhcbInfo Info>
+  class SevGhcbMsrSimpleRequest
+  {
+    uint64_t content;
+
+  public:
+    SevGhcbMsrSimpleRequest(uint64_t input) : content(Info | (input << 12)) {}
```
Comment (mjp41) [inline guest-verona-rt/arch/x86_64/inc/sev_msr.h]: Asserts here to check that packing actually works, i.e. input doesn't have leading 16bits set, and Info is less than 2^12?

**Review by mjp41 — APPROVED** [review]
As far as I understand looks good.  A couple of minors bits

File: guest-verona-rt/arch/x86_64/hv_sev.cc (hunk 6 above)
Comment (ihaller) [reply guest-verona-rt/arch/x86_64/hv_sev.cc:530]: It is not ideal that other threads continue running, but there is no way for us to force other threads to exit, not even if we triple-fault.

File: guest-verona-rt/arch/x86_64/hv_sev.cc (hunk 5 above)
Comment (ihaller) [reply guest-verona-rt/arch/x86_64/hv_sev.cc:520]: Yes, we need to do more testing over different guests to make sure that the lazy acceptance works well. I will do it in a follow-up.

File: guest-verona-rt/arch/x86_64/hv_sev.cc (hunk 3 above)
Comment (ihaller) [reply guest-verona-rt/arch/x86_64/hv_sev.cc]: You are right, I should initialize `last_entry_end_address` to `snmalloc::address_cast(heap_start)`

File: guest-verona-rt/arch/x86_64/hv_sev.cc (hunk 2 above)
Comment (ihaller) [reply guest-verona-rt/arch/x86_64/hv_sev.cc]: I wasn't sure if I needed to use `last_entry_end_address` later on when building out the checks, but now that it is more final your proposal makes sense.

File: guest-verona-rt/arch/x86_64/inc/sev_msr.h (hunk 7 above)
Comment (ihaller) [reply guest-verona-rt/arch/x86_64/inc/sev_msr.h]: Info is typed to the enum, but I added an assert for input.

**Comment by ihaller** [comment]
Updated the heap acceptance logic based on the feedback and added the assert.
